File: swift/Sources/MemCardApp/Model/Basket.swift
import Foundation
import MemCardKit

/// Корзина: то, из чего соберётся карта. Копится по дороге, из любого места.
@MainActor
@Observable
final class Basket {
    private(set) var items: [LibraryItem] = []
    /// Что не влезло или было отброшено - показываем, а не молчим.
    private(set) var note: String?

    var used: Int { items.reduce(0) { $0 + $1.blocks } }
    var free: Int { PSX.slots - used }
    var isEmpty: Bool { items.isEmpty }

    func contains(_ item: LibraryItem) -> Bool {
        items.contains { $0.fingerprint == item.fingerprint }
    }

    func toggle(_ item: LibraryItem) {
        if let index = items.firstIndex(where: { $0.fingerprint == item.fingerprint }) {
            items.remove(at: index)
            note = nil
            return
        }
        add(item)
    }

    func add(_ item: LibraryItem) {
        guard !contains(item) else { return }
        guard item.blocks <= free else {
            note = L.t("{0} takes {1} {2}, only {3} free", item.title, item.blocks, L.plural("block", item.blocks), free)
            return
        }
        // Игра находит сейв по имени, поэтому двух одинаковых имён на карте
        // быть не должно. Разные сейвы под одним именем выбирать за
        // пользователя нельзя - говорим и не берём.
        if let clash = items.first(where: { $0.save.name == item.save.name }) {
            note = L.t("\"{0}\" is already in the basket — from {1}. The game tells saves apart by name; two identical names cannot share a card", item.save.name, clash.title)
            return
        }
        items.append(item)
        note = nil
    }

    func remove(_ item: LibraryItem) {
        items.removeAll { $0.fingerprint == item.fingerprint }
        note = nil
    }

    func clear() {
        items = []
        note = nil
    }

    /// Раскладка по пятнадцати слотам: у многоблочного сейва видна вся цепочка.
    struct Cell: Identifiable {
        let id: Int
        let item: LibraryItem?
        let isContinuation: Bool
    }

    var layout: [Cell] {
        var cells: [Cell] = []
        for item in items {
            for position in 0..<item.blocks {
                cells.append(Cell(id: cells.count, item: item,
                                  isContinuation: position > 0))
            }
        }
        while cells.count < PSX.slots {
            cells.append(Cell(id: cells.count, item: nil, isContinuation: false))
        }
        return cells
    }

    func build() throws -> CardBuilder.Result {
        try CardBuilder.build(items.map(\.save))
    }
}

File: qt/nsm/basket.py
"""Корзина: то, из чего соберётся карта. Копится по дороге, из любого места."""

import os
import sys

HERE = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
TOOLS = os.path.join(HERE, "tools")
if TOOLS not in sys.path:
    sys.path.insert(0, TOOLS)

import psxbuild  # noqa: E402

SLOTS = 15


class Basket:
    def __init__(self):
        self.items = []
        self.note = None

    @property
    def used(self):
        return sum(item.blocks for item in self.items)

    @property
    def free(self):
        return SLOTS - self.used

    def contains(self, item):
        return any(x.fingerprint == item.fingerprint for x in self.items)

    def toggle(self, item):
        if self.contains(item):
            self.items = [x for x in self.items
                          if x.fingerprint != item.fingerprint]
            self.note = None
            return
        self.add(item)

    def add(self, item):
        if self.contains(item):
            return
        if item.blocks > self.free:
            self.note = (f"{item.title} занимает {item.blocks} "
                         f"{blocks_word(item.blocks)}, а свободно {self.free}")
            return
        # Игра находит сейв по имени, поэтому двух одинаковых имён на
        # карте быть не должно. Выбирать за пользователя нельзя - говорим.
        name = bytes(item.frame[10:30])
        clash = next((x for x in self.items
                      if bytes(x.frame[10:30]) == name), None)
        if clash:
            self.note = (f"«{name.split(b'\\x00')[0].decode('ascii', 'replace')}» "
                         f"уже в корзине — из {clash.title}. Игра различает "
                         f"сейвы по имени, двух одинаковых быть не может")
            return
        self.items.append(item)
        self.note = None

    def clear(self):
        self.items = []
        self.note = None

    def layout(self):
        """Раскладка по пятнадцати слотам: у многоблочного видна цепочка."""
        cells = []
        for item in self.items:
            for position in range(item.blocks):
                cells.append((item, position > 0))
        while len(cells) < SLOTS:
            cells.append((None, False))
        return cells

    def build(self):
        """Собирает образ карты из выбранного."""
        entries = []
        for item in self.items:
            blocks = [item.block[at:at + psxbuild.psxid.BLOCK]
                      for at in range(0, len(item.block), psxbuild.psxid.BLOCK)]
            entries.append({"name": bytes(item.frame[10:30]),
                            "blocks": blocks, "where": item.where})
        return psxbuild.build(entries)


def saves_word(count):
    tail = count % 100
    if 11 <= tail <= 14:
        return "сейвов"
    return {1: "сейв", 2: "сейва", 3: "сейва", 4: "сейва"}.get(count % 10, "сейвов")


def blocks_word(count):
    tail = count % 100
    if 11 <= tail <= 14:
        return "блоков"
    return {1: "блок", 2: "блока", 3: "блока", 4: "блока"}.get(count % 10, "блоков")
